src/signature.sh

```diff
@@ -0,0 +1,402 @@
+declare -gr email_regex='[A-Za-z0-9_\.-]+@[A-Za-z0-9_-]+(\.[A-Za-z0-9]+)+'
+
+# This function performs operations over trailers in
+# either patches or commits. It checks if given argument
+# is a valid commit reference or patch path and uses the
+# correct command to perform the task.
+# If that's not the case, a warning message will tell
+# the user this argument was ignored.
+#
+# Also, if no operation option is given, then an error message
+# followed by a helper message is printed to the user.
+#
+# @patch_or_sha_args Holds either patch paths or commit references.
+function signature_main()
+{
+  local patch_or_sha_args
+  local flag
+
+  if [[ "$1" =~ -h|--help ]]; then
+    signature_help "$1"
+    exit 0
+  fi
+
+  parse_signature_options "$@"
+  if [[ "$?" -gt 0 ]]; then
+    complain "${options_values['ERROR']}"
+    return 22 # EINVAL
+  fi
+
+  [[ -n "${options_values['VERBOSE']}" ]] && flag='VERBOSE'
+  flag=${flag:-'SILENT'}
+
+  read -ra patch_or_sha_args <<< "${options_values['PATCH_OR_SHA']}"
+
+  if [[ "${#all_trailers[@]}" -eq 0 ]]; then
+    complain 'An option is required to use this command.'
+    signature_help
+    return 22 # EINVAL
+  fi
+
+  for arg in "${patch_or_sha_args[@]}"; do
+    write_all_trailers "$arg" "$flag"
+  done
+}
+
+# This function validates the signature. If the passed signature
+# does not follow the format 'NAME <EMAIL>' with a valid EMAIL, then
+# it returns an error.
+#
+# @signature Holds the string that is validated
+#
+# Return:
+# Returns 0 if signature is valid; returns 22 otherwise.
+function is_valid_signature()
+{
+  local signature="$1"
+  local signature_regex='^[^<]+ <'"$email_regex"'>$'
+
+  if ! [[ "$signature" =~ $signature_regex ]]; then
+    return 22 # EINVAL
+  fi
+}
+
+# This function parses and adds a new trailer line into
+# @all_trailers that will be properly written later.
+# It attempts to use the user's name and email configured
+# if a signature is not passed as argument and gives
+# an error if they are not properly set with git config.
+#
+# @keyword Holds a string that is a keyword for a specific operation.
+# @signature Holds a string like 'NAME <EMAIL>' defining the signature.
+#
+# Return:
+# In case of successful return 0 adding the parsed operation;
+# It returns 1 if either user.name or user.email are not configured properly;
+# It returns 22 if the signature does not follow 'NAME <EMAIL>' format;
+function parse_and_add_trailer()
+{
+  local keyword="$1"
+  local signature="$2"
+  local formated_output
+  local trailer
+
+  trailer="${TRAILER_TAGS[$keyword]} "
+
+  # Use default from git config if no argument was given
+  if [[ ! "$signature" ]]; then
+    formated_output="$(format_name_email_from_user)"
+    if [[ "$?" -gt 0 ]]; then
+      return 1
+    fi
+    trailer+="$formated_output"
+  else
+    signature="$(str_strip "$signature")"
+    if [[ "$keyword" != 'FIXES' ]] && ! is_valid_signature "$signature"; then
```

Comment:
Wait... why is FIXES a factor here but not in the above ones? I did not get it.

Reply:
The FIXES factor is not necessary above because the `--add-fixes` options demands an argument, which means that it's not possible (or at least it shouldn't) for the user to call this option with empty content. When I wrote this code I assumed that if `"$signature"` is not given then the caller is not trying to parse a `Fixes` tag line, unless the developer working on this feature (in that case, me) messed up and forgot to pass the second parameter.

In this line we need the FIXES factor because this tag's content follows a different format and it was already validated and parsed previously in `format_fixes_message`.

I'm open to suggestions on how to improve this code's flow though. I agree it might be confusing.